src/index.js

```diff
@@ -29,369 +23,342 @@ class IPLDResolver {
+   * @returns {this}
+   */
+  removeFormat (codec) {
+    if (this.resolvers[codec]) {
+      delete this.resolvers[codec]
     }
 
-    if (path === '' || !path) {
-      return this._get(cid, (err, node) => {
-        if (err) {
-          return callback(err)
-        }
-        callback(null, {
-          value: node,
-          remainderPath: '',
-          cid
-        })
-      })
+    return this
+  }
+
+  /**
+   * Retrieves IPLD Nodes along the `path` that is rooted at `cid`.
+   *
+   * @param {CID} cid - the CID the resolving starts.
+   * @param {string} path - the path that should be resolved.
+   * @returns {Iterable.<Promise.<{remainderPath: string, value}>>} - Returns an async iterator of all the IPLD Nodes that were traversed during the path resolving. Every element is an object with these fields:
+   *   - `remainderPath`: the part of the path that wasn’t resolved yet.
+   *   - `value`: the value where the resolved path points to. If further traversing is possible, then the value is a CID object linking to another IPLD Node. If it was possible to fully resolve the path, value is the value the path points to. So if you need the CID of the IPLD Node you’re currently at, just take the value of the previously returned IPLD Node.
+   */
+  resolve (cid, path) {
+    if (!CID.isCID(cid)) {
+      throw new Error('`cid` argument must be a CID')
+    }
+    if (typeof path !== 'string') {
+      throw new Error('`path` argument must be a string')
     }
 
-    let value
-
-    doUntil(
-      (cb) => {
-        this._getFormat(cid.codec, (err, format) => {
-          if (err) return cb(err)
-
-          // get block
-          // use local resolver
-          // update path value
-          this.bs.get(cid, (err, block) => {
-            if (err) {
-              return cb(err)
-            }
-
-            format.resolver.resolve(block.data, path, (err, result) => {
-              if (err) {
-                return cb(err)
-              }
-              value = result.value
-              path = result.remainderPath
-              cb()
-            })
-          })
-        })
-      },
-      () => {
-        const endReached = !path || path === '' || path === '/'
-        const isTerminal = value && !IPLDResolver._maybeCID(value)
-
-        if ((endReached && isTerminal) || options.localResolve) {
-          cid = IPLDResolver._maybeCID(value) || cid
-
-          return true
-        } else {
-          value = IPLDResolver._maybeCID(value)
-          // continue traversing
-          if (value) {
-            cid = value
+    const generator = async function * () {
+      // End iteration if there isn't a CID to follow anymore
+      while (cid !== null) {
+        const format = await this._getFormat(cid.codec)
+
+        // get block
+        // use local resolver
+        // update path value
+        const block = await promisify(this.bs.get.bind(this.bs))(cid)
```

Comment:
Why not do the promisification of the blockstore methods in the constructor instead of repeatedly during loops (e.g. here, during `.get`, `.remove`, `.tree`, etc)?

e.g.

```javascript
class IPLDResolver {
  constructor (opts) {
    this.bs = {
       get: promisify(opts.bs.get.bind(opts.bs)),
       put: promisify(opts.bs.put.bind(opts.bs)),
       delete: promisify(opts.bs.delete.bind(opts.bs)),
       // ...etc
    }

    // ...etc
  }
}
```

Alternatively use something like [Bluebird.promisifyAll](http://bluebirdjs.com/docs/api/promisification.html) until the lower level libraries are promise-aware?